src/buffer_field_api.h

```diff
@@ -0,0 +1,311 @@
+#ifndef __dispatch_buffer_field_api_h__
+#define __dispatch_buffer_field_api_h__ 1
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/** @file
+ * Inline API for common operations on buffer_fields
+ * @internal
+ * @{
+ */
+
+#include "qpid/dispatch/buffer_field.h"
+#include "qpid/dispatch/iterator.h"
+
+/* qd_buffer_field_memcpy
+ *
+ * Copy up to n octets from bfield to dest, advance bfield by the number of
+ * octets copied
+ *
+ * @return total of octets copied - may be < n if len(bfield) < n
+ */
+static inline size_t qd_buffer_field_memcpy(qd_buffer_field_t *bfield, uint8_t *dest, size_t n)
+{
+    const uint8_t *start = dest;
+    size_t count = MIN(n, bfield->length);
+    if (bfield->buffer) {
+        while (count) {
+            size_t avail = qd_buffer_cursor(bfield->buffer) - bfield->cursor;
+            if (count < avail) {
+                // fastpath: no need to adjust buffer pointers
+                memcpy(dest, bfield->cursor, count);
+                dest += count;
+                bfield->cursor += count;
+                bfield->length -= count;
+                return dest - start;
+            }
+
+            // count is >= what is available in the current buffer
+            memcpy(dest, bfield->cursor, avail);
+            dest += avail;
+            count -= avail;
+            bfield->length -= avail;
+            bfield->cursor += avail;
+            if (bfield->length) {
+                bfield->buffer = DEQ_NEXT(bfield->buffer);
+                if (bfield->buffer) {
+                    bfield->cursor = (const uint8_t *)qd_buffer_base(bfield->buffer);
+                } else {
+                    // DISPATCH-1394: field is truncated (length is inaccurate!)
+                    bfield->length = 0;
+                    count = 0;
+                    assert(false);  // TODO(KAG): is this fixed?
+                }
+            }
+        } while (count);
+
+        return dest - start;
+
+    } else {    // string/binary data
```

Comment:
Here's the problem (which is totally non-obvious) - qd_iterator_t (from which the qd_buffer_field_t api was copied) supports buffer fields that don't actually have an associated qd_buffer_list_t!  When an iterator is created via qd_iterator_string() or qd_iterator_binary() there's no qd_buffer_t involved at all.

I didn't realize that until parse tests started failing since the code was attempting to parse out an iterator that was created on a string/binary non-buffer.  So I had to carry this over to minimize the patch.

I agree this is non-obvious and messy.

To remove the non-buffer code from the qd_buffer_field_t will require modifying parse() to check the iterator parameter for a buffer.  If not present parse() _could_ copy the iterator's content into a temp buffer list.  That may not be as bad as it sounds if it only occurs for non-message parsing - I'll take a look at that.